CSS Nesting: support top-level & selector

mdubet opened the pull request:
#### 20329b62061b40d5a423a1d75b67779945b84729
<pre>
CSS Nesting: support top-level &amp; selector
<a href="https://bugs.webkit.org/show_bug.cgi?id=250379">https://bugs.webkit.org/show_bug.cgi?id=250379</a>
rdar://104073866

Reviewed by Antti Koivisto.

Replaces any top-level &amp; selector by the :scope pseudo class.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/invalidation-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/supports-rule-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/supports-rule-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/supports-rule.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/top-level-is-scope-expected.txt:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::replaceNestingParentByPseudoClassScope):
(WebCore::CSSSelector::replaceNestingParentByNotAll): Deleted.
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/CSSStyleRule.cpp:
(WebCore::CSSStyleRule::setSelectorText):
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::CSSSelectorParser::consumeSimpleSelector):
(WebCore::CSSSelectorParser::resolveNestingParent):
* Source/WebCore/dom/SelectorQuery.cpp:
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::populateStyleRuleResolvedSelectorList):

Canonical link: <a href="https://commits.webkit.org/261739@main">https://commits.webkit.org/261739@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/66d6b83478f68c4cdc642e82540d6ab23bc62968

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/112549 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/21699 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/1204 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/4323 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/121105 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/23037 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/12860 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/5464 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/118316 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/17116 "5 failures") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/100318 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/105625 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/99061 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/88/builds/875 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/46132 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/100853 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/14037 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/89/builds/900 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/95238 "13 api tests failed or timed out") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/12138 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/14719 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/10277 "2 failures") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/102329 "Built successfully") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/20058 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52905 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/31965 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/74/builds/8148 "The change is no longer eligible for processing.") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/16555 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/110368 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4489 "Built successfully and passed tests") | | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/27265 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/css/CSSSelectorList.cpp:
```diff
@@ -193,4 +193,52 @@ bool CSSSelectorList::hasExplicitNestingParent() const
+    }
+
+    auto final = CSSSelectorList { WTFMove(result) };
+    return final;
```
`final` is a keyword so this looks funny. Also you can just return without a local.

mdubet replied:
Indeed, this is a leftover of debugging it

anttijk commented on Source/WebCore/css/CSSSelectorList.cpp:
```diff
@@ -193,4 +193,52 @@ bool CSSSelectorList::hasExplicitNestingParent() const
+                else
+                    parentIsSelector.setNotLastInSelectorList();
+
+                auto uniqueParentIsSelector = makeUnique<CSSParserSelector>(parentIsSelector);
```
not sure what the word unique in variable name signifies.

mdubet replied:
Renamed

anttijk commented on Source/WebCore/css/CSSSelectorList.cpp:
```diff
@@ -193,4 +193,52 @@ bool CSSSelectorList::hasExplicitNestingParent() const
+    while (selector) {
+        if (selector->hasExplicitNestingParent()) {
+            if (parentResolvedSelectorList)
+                const_cast<CSSSelector*>(selector)->resolveNestingParentSelectors(*parentResolvedSelectorList);
```
It is unfortunate that CSSSelector is mutated like this.

anttijk commented on Source/WebCore/css/CSSSelectorList.h:
```diff
@@ -68,6 +68,7 @@ class CSSSelectorList {
     unsigned listSize() const;
 
     CSSSelectorList& operator=(CSSSelectorList&&) = default;
+    static CSSSelectorList resolveNestingParent(const CSSSelectorList& nestedSelectorList, const CSSSelectorList* parentResolvedSelectorList);
```
Why can't this stay in `CSSSelectorParser`?

mdubet replied:
It's not really related to the selector parser, it makes more sense to have it in `CSSSelectorList`.

anttijk replied:
Neither `CSSSelectorList` nor `CSSSelector` should be in the business of building themselves. I'd like a factoring where all building and mutation code is separate. It is true that`CSSSelectorParser` may not be a great place for this (we might want `CSSSelectorBuilder` or something) but I don't see why this patch needs to move the function.

mdubet replied:
The refactoring has been moved out of this PR

mdubet committed:
CSS Nesting: support top-level & selector
https://bugs.webkit.org/show_bug.cgi?id=250379
rdar://104073866

Reviewed by Antti Koivisto.

Replaces any top-level & selector by the :scope pseudo class.

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/invalidation-004-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/supports-rule-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/supports-rule-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/supports-rule.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/top-level-is-scope-expected.txt:
* Source/WebCore/css/CSSSelector.cpp:
(WebCore::CSSSelector::replaceNestingParentByPseudoClassScope):
(WebCore::CSSSelector::replaceNestingParentByNotAll): Deleted.
* Source/WebCore/css/CSSSelector.h:
* Source/WebCore/css/CSSStyleRule.cpp:
(WebCore::CSSStyleRule::setSelectorText):
* Source/WebCore/css/parser/CSSSelectorParser.cpp:
(WebCore::CSSSelectorParser::consumeSimpleSelector):
(WebCore::CSSSelectorParser::resolveNestingParent):
* Source/WebCore/dom/SelectorQuery.cpp:
* Source/WebCore/style/RuleSetBuilder.cpp:
(WebCore::Style::RuleSetBuilder::populateStyleRuleResolvedSelectorList):

Canonical link: https://commits.webkit.org/261739@main
